tests/AstrolabeWorkloadExecutor/Program.cs

```diff
@@ -0,0 +1,114 @@
+        {
+            var connectionString = args[0];
+            var driverWorkload = BsonDocument.Parse(args[1]);
+
+            var cancellationTokenSource = new CancellationTokenSource();
+            ConsoleCancelEventHandler cancelHandler = (o, e) => HandleCancel(e, cancellationTokenSource);
+
+            var resultsDir = Environment.GetEnvironmentVariable("RESULTS_DIR");
+            var resultsPath = resultsDir == null ? "results.json" : Path.Combine(resultsDir, "results.json");
+            Console.WriteLine($"dotnet main> Results will be written to {resultsPath}..." );
+
+            try
+            {
+                Console.CancelKeyPress += cancelHandler;
+
+                Console.WriteLine("dotnet main> Starting workload executor...");
+
+                if (!bool.TryParse(Environment.GetEnvironmentVariable("ASYNC"), out bool async))
+                {
+                    async = true;
+                }
+
+                ExecuteWorkload(connectionString, driverWorkload, async, cancellationTokenSource.Token);
+            }
+            finally
+            {
+                Console.CancelKeyPress -= cancelHandler;
+                Console.WriteLine("dotnet main finally> Writing final results file");
+                var resultsJson = ConvertResultsToJson();
+                Console.WriteLine(resultsJson);
+#if NETCOREAPP2_1
+                File.WriteAllTextAsync(resultsPath, resultsJson).Wait();
+#else
+                File.WriteAllText(resultsPath, resultsJson);
+#endif
+            }
+        }
+
+        static string ConvertResultsToJson()
+        {
+                var resultsJson =
+                @"{ " +
+                $"  \"numErrors\" : {Interlocked.Read(ref __numberOfOperationErrors)}, " +
+                $"  \"numFailures\" : {Interlocked.Read(ref __numberOfFailedOperations)}, " +
+                $"  \"numSuccesses\" : {Interlocked.Read(ref __numberOfSuccessfulOperations)}  " +
+                @"} ";
+
+                return resultsJson;
+        }
+
+        static void ExecuteWorkload(string connectionString, BsonDocument driverWorkload, bool async, CancellationToken cancellationToken)
+        {
+            Environment.SetEnvironmentVariable("MONGODB_URI", connectionString);
+
+            var testRunner = new AstrolabeTestRunner(
+                incrementOperationSuccesses: () => Interlocked.Increment(ref __numberOfSuccessfulOperations),
+                incrementOperationErrors: () => Interlocked.Increment(ref __numberOfOperationErrors),
+                incrementOperationFailures: () => Interlocked.Increment(ref __numberOfFailedOperations),
+                cancellationToken: cancellationToken);
+            var factory = new AstrolabeTestRunner.TestCaseFactory();
+            var testCase = factory.CreateTestCase(driverWorkload, async);
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                testRunner.Run(testCase);
```

Comment:
Note the sync call to `Run` even if async is true. We're not really doing true async testing.

Then again... I'm not sure we even need to do async testing.